File: studio/component_html.py
"""Extract/replace inner HTML for data-section marked component roots."""

import re

_SECTION_OPEN_RE = re.compile(
    r'<(?P<tag>\w+)(?P<attrs>[^>]*\bdata-section="(?P<type>[^"]+)"[^>]*)>',
    re.I | re.S,
)


def _tag_re(tag):
    return re.compile(rf"<(?P<slash>/?){re.escape(tag)}\b[^>]*>", re.I)


def _sections(html):
    """Yield (type, open_start, inner_start, inner_end, end) per section root.

    Depth-aware: a non-greedy regex closes on the first nested </div>, which
    truncated every div-rooted component (stats, features, about...).
    """
    for m in _SECTION_OPEN_RE.finditer(html):
        depth = 1
        for t in _tag_re(m.group("tag")).finditer(html, m.end()):
            depth += -1 if t.group("slash") else 1
            if depth == 0:
                yield m.group("type"), m.start(), m.end(), t.start(), t.end()
                break


def list_section_types(html):
    return [s[0] for s in _sections(html)]


def extract_section_inner(html, section_type):
    for type_, _, inner_start, inner_end, _ in _sections(html):
        if type_ == section_type:
            return html[inner_start:inner_end]
    return None


def replace_section_inner(html, section_type, new_inner):
    from studio import draft

    new_inner = draft._strip_markdown_fences(new_inner)
    for type_, _, inner_start, inner_end, _ in _sections(html):
        if type_ == section_type:
            return html[:inner_start] + new_inner + html[inner_end:]
    return html


def validate_section_root(opening_tag_html, section_type, *, expected_scope_prefix):
    if f'data-section="{section_type}"' not in opening_tag_html:
        return False, "missing_data_section"
    if expected_scope_prefix and expected_scope_prefix not in opening_tag_html:
        return False, "missing_scope_class"
    return True, ""

File: src/server/componentHtml.ts
/** Extract/replace inner HTML for data-section marked component roots. */

const SECTION_OPEN_RE =
  /<(\w+)([^>]*\bdata-section="([^"]+)"[^>]*)>/gi;

function tagRe(tag: string): RegExp {
  return new RegExp(`<(/?)${tag.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}\\b[^>]*>`, "gi");
}

type SectionSpan = [string, number, number, number, number];

function sections(html: string): SectionSpan[] {
  const out: SectionSpan[] = [];
  for (const m of html.matchAll(SECTION_OPEN_RE)) {
    const tag = m[1]!;
    const type = m[3]!;
    const re = tagRe(tag);
    re.lastIndex = m.index! + m[0].length;
    let depth = 1;
    for (const t of html.matchAll(re)) {
      depth += t[1] ? -1 : 1;
      if (depth === 0) {
        out.push([type, m.index!, m.index! + m[0].length, t.index!, t.index! + t[0].length]);
        break;
      }
    }
  }
  return out;
}

export function extractSectionInner(html: string, sectionType: string): string | null {
  for (const [type, , innerStart, innerEnd] of sections(html)) {
    if (type === sectionType) return html.slice(innerStart, innerEnd);
  }
  return null;
}

export function replaceSectionInner(html: string, sectionType: string, newInner: string): string {
  const clean = newInner.replace(/^```(?:html|htm|xml|json)?\s*|\s*```$/gim, "");
  for (const [type, , innerStart, innerEnd] of sections(html)) {
    if (type === sectionType) return html.slice(0, innerStart) + clean + html.slice(innerEnd);
  }
  return html;
}
